### ISSUE TITLE
ABCL backend possibly broken by commit 57b6d3fd2d4634313cd23c83a3fbd7fa53ad55a4

### ISSUE THREAD

**Issue by zellerin** [joaotavora/sly#425]
Hi,

ABCL backend does not work for me now, I believe commit 57b6d3fd2d4634313cd23c83a3fbd7fa53ad55a4 has caused this by
- having `slynk/abcl` (formerly `slynk-abcl`) depend on non-existent `slynk/backend` (`slynk-backend` exists), and 
- using `slynk::` symbols, when slynk package is defined only later in slynk.lisp (see the .asd file)
- similar issues for `slynk/match` - both rename and order of loading.

Probably related to failure in https://common-lisp.net/project/cl-test-grid/library/sly.html.

**Comment by oolonc** [comment]
I believe I am running into this as well. Trying to loadk `slynk` gives me

```
(ql:quickload :slynk)
To load "slynk":
  Load 1 ASDF system:
    slynk
; Loading "slynk"
.............................
#<THREAD "interpreter" {1D7FD8D2}>: Debugger invoked on condition of type READER-ERROR
  The package "SLYNK/BACKEND" can't be found.
Restarts:
  0: RETRY                         Retry compiling #<ASDF/LISP-ACTION:CL-SOURCE-FILE "slynk" "backend" "abcl">.
  1: ACCEPT                        Continue, treating compiling #<ASDF/LISP-ACTION:CL-SOURCE-FILE "slynk" "backend" "abcl"> as having been successful.
  2: RETRY                         Retry ASDF operation.
  3: CLEAR-CONFIGURATION-AND-RETRY Retry ASDF operation after resetting the configuration.
  4: RETRY                         Retry ASDF operation.
  5: CLEAR-CONFIGURATION-AND-RETRY Retry ASDF operation after resetting the configuration.
  6: ABORT                         Give up on "slynk"
  7: REGISTER-LOCAL-PROJECTS       Register local projects and try again.
  8: TOP-LEVEL                     Return to top level.
[1] KEYWORD(2):
```

**Comment by joaotavora** [comment]
Yes, this is in the backlog.  it needs someone to go through the file slowly and fix it :-)  :+1: 

@zellerin made some good analysis that should be easy to pick up.

**Comment by zellerin** [comment]
Short term solution to just make it work is to remove the discussed commit and all newer commits prefixed abcl:.

I am now again on the machine with Java, so I tried  this and abcl loads. As @joaotavora  said, the proper approach is to go slowly through it and test with abcl as it goes. Apologies, I use abcl only as last resort, so I am not much willing to commit myself to this.

Incidentally, this seems to indicate that abcl backend is not much used.

**Comment by gwbrown** [comment]
I spent a couple hours looking at this and it's unclear how one would fix this in a way that would be acceptable. ABCL is an unusual implementation in that there are lots of translations that have to happen between Common Lisp-land and Java-land, and the backend is doing a lot of work to make that ergonomic. In Slime the backend can easily call into functions in `SWANK`, presumably because of whatever their custom loader thing does, but in Sly the loading order is set by the ASD.

There are  a few options I can think of:
1. The (apparently unacceptable) option in #449, of breaking out a separate file to define some things out of order.
2. Using an evil hack to funcall the symbol dynamically, something like:
```lisp
(defun evil-hack (function &rest args)
  (apply (read-from-string function) args))
;; (slynk:eval-in-emacs form t) becomes
(evil-hack "slynk:eval-in-emacs" form t)
```
This works mostly, except for a call to `swank::lcons*`, which is a macro and cannot be addressed with this strategy. I haven't had time to dig in to find out if there would be a way to rewrite it without this macro, but it might lead to a big diff (which is to be avoided).
3. Rip out a bunch of functionality entirely. Which would still lead to a big diff.

**Comment by gwbrown** [comment]
Actually the lcons* is a bug in Slime too, as the macro isn't defined when the ABCL backend is compiled, and I was getting errors trying to inspect strings while using it earlier. So whatever the solution is should probably be contributed there too.

**Comment by joaotavora** [comment]
@gwbrown thanks for the analysis.  There are, you'll see some pretty big evil hacks in SLY's backend code.  I wonder if SLIME's ACBL backend isn't using them.

If you're using SLY and ABCL and you want to maintain SLY's `abcl.lisp` backend (and ensure it works for say, the most recent/stable version of ABCL), you can do the refactorings you think are worthy.  If you follow the contributing guidelines (I'll explain what those are, there aren't that many), I can give you commit rights.  That work will be very appreciated.

**Comment by gwbrown** [comment]
Thanks; I'm not terribly familiar with the Sly/SLIME code except for the bits required to do basic troubleshooting here. The offer to give me commit rights is very generous, and I *might* take you up on that - I'm not using ABCL for a particular project at the moment, but I do have reason to use it semi-frequently for experiments and such. I'll try to dig into this more soon (though with holidays and such it might be a minute).

**Comment by joaotavora** [comment]
> The offer to give me commit rights is very generous, and I might take you up on that 

It's more of an exchange rather than straight-up offer ;-) you maintain code you get the commit rights to do it :-D
